build/release/Makefile

```diff
@@ -0,0 +1,132 @@
+# Copyright 2016 The Rook Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+all: build
+
+include ../makelib/common.mk
+
+# ====================================================================================
+# Options
+
+CHANNEL ?= master
+ifeq ($(filter master alpha beta stable,$(CHANNEL)),)
+$(error invalid channel $(CHANNEL))
+endif
+
+BIN_DIR ?= $(abspath ../../bin)
+PLATFORMS ?= $(ALL_PLATFORMS)
+FLAVORS ?= binaries images
+TEMP_DIR := $(shell mktemp -d)
+
+DOCKER_REGISTRY ?= rook
+QUAY_REGISTRY ?= quay.io/rook
+REGISTRIES ?= $(DOCKER_REGISTRY) $(QUAY_REGISTRY)
+IMAGES ?= rook toolbox
+IMAGE_ARCHS := $(subst linux_,,$(filter linux_%,$(PLATFORMS)))
+IMAGE_PLATFORMS := $(subst _,/,$(subst $(SPACE),$(COMMA),$(filter linux_%,$(PLATFORMS))))
+
+S3_BUCKET ?= rook-release
+S3_CP := aws s3 cp --only-show-errors
+S3_SYNC := aws s3 sync --only-show-errors --delete
+S3_CURRENT_BUCKET := s3://$(S3_BUCKET)/$(CHANNEL)/$(VERSION)
+
+# ====================================================================================
+# tools
+
+MANIFEST_TOOL_VERSION=v0.5.0
+MANIFEST_TOOL := $(TOOLS_HOST_DIR)/manifest-tool-$(MANIFEST_TOOL_VERSION)
+
+$(MANIFEST_TOOL):
+	@echo === installing manifest-tool
+	@curl -sL https://github.com/estesp/manifest-tool/releases/download/v0.5.0/manifest-tool-$(GOHOSTOS)-$(GOHOSTARCH) > $@
+	@chmod +x $@
+
+# ====================================================================================
+# Targets
+
+build: $(addprefix build.,$(FLAVORS)) ;
+publish: $(addprefix publish.,$(FLAVORS)) ;
+promote: $(addprefix promote.,$(FLAVORS)) ;
+clean: $(addprefix clean.,$(FLAVORS)) ;
+
+# catch all for unimplemented targets / flavors
+%: ; @:
+
+# ====================================================================================
+# binary files
+
+publish.binaries.%:
+	@$(S3_CP) $(BIN_DIR)/$*/rookctl$(if $(filter windows_%,$*),.exe) $(S3_CURRENT_BUCKET)/$*/rookctl$(if $(filter windows_%,$*),.exe)
+publish.binaries: $(addprefix publish.binaries.,$(PLATFORMS))
+	@echo "$(VERSION)" > $(TEMP_DIR)/version;
+	@$(S3_CP) "$(TEMP_DIR)/version" $(S3_CURRENT_BUCKET)/version
+promote.binaries:
+	@$(S3_SYNC) s3://$(S3_BUCKET)/master/$(VERSION) s3://$(S3_BUCKET)/$(CHANNEL)/current
+
+# ====================================================================================
+# images
+
+define repo.targets
```

Comment:
is it possible (and useful) to add comment to this gnarly target generation that tell the reader what 1,2,3 are? someone experienced with makefiles could tell that they are registry, image and platform from looking at the eval call on line 98, but a comment may be useful for those less experienced with make.